### PR TITLE
Geo/check number of tables phreatic multi line

### PR DESCRIPTION
**Pull request by WPK4FEM**
**📝 Description**
To make sure that the number of tables matches the number of points of the multlinear phreatic line input a check has been added.

**🆕 Changelog**
- Check added
- Unit test added
- Some scrambled and redundant project parameters syntax corrected.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp [hunk 1]
```diff
@@ -51,12 +51,12 @@ void ApplyConstantPhreaticMultiLinePressureProcess::InitializeParameters(Paramet
                 "is_seepage": false,
                 "gravity_direction": 1,
                 "out_of_plane_direction": 2,
-                "x_coordinates":           [0.0,1.0],
-                "y_coordinates":           [1.0,0.5],
-				"z_coordinates":           [0.0,0.0],
+                "x_coordinates":           [0.0, 1.0],
+                "y_coordinates":           [1.0, 0.5],
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp:55]: It seems quite random that the defaults have non-zero values here, but that might be a larger discussion

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp [hunk 2]
```diff
@@ -129,14 +129,15 @@ KRATOS_TEST_CASE_IN_SUITE(ApplyConstantPhreaticMultiLinePressureProcessDoesNotTh
                 "y_coordinates": [0.0, 1.0],
                 "z_coordinates": [0.0, 0.0],
                 "gravity_direction": 1,
-                "out_of_plane_direction": 2
+                "out_of_plane_direction": 2,
+                "table": [0, 0]
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp]: I would expect the 'constant' process doesn't need a table here

File: applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp [hunk 3]
```diff
@@ -51,12 +51,12 @@ void ApplyConstantPhreaticMultiLinePressureProcess::InitializeParameters(Paramet
...
+                "z_coordinates":           [0.0, 0.0],
                 "specific_weight" : 10000.0,
                 "pressure_tension_cut_off" : 0.0,
-                "table" : [0,1]
+                "table" : [0, 0]
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp:59]: Would it be possible to have this default only in the table version of this process?

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_phreatic_multi_line_pressure_table_process.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                "z_coordinates": [0.0, 0.0],
+                "gravity_direction": 1,
+                "out_of_plane_direction": 2,
+                "table": [1, 2, 3, 4]
+            }  )"};
+
+    KRATOS_EXPECT_EXCEPTION_IS_THROWN(
+        (ApplyPhreaticMultiLinePressureTableProcess{r_model_part, test_parameters}),
+        "Got 2 coordinates and 4 table references. The number of coordinates and table references should be equal.")
+}
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_phreatic_multi_line_pressure_table_process.cpp]: Let's also add a test which doesn't throw here

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp [hunk 5]
```diff
@@ -129,14 +129,15 @@ KRATOS_TEST_CASE_IN_SUITE(ApplyConstantPhreaticMultiLinePressureProcessDoesNotTh
...
             }  )"};
 
     KRATOS_EXPECT_TRUE(CanCreateInstanceOfApplyConstantPhreaticMultiLinePressureProcessWithoutFailure(
         r_model_part, test_parameters))
 
     test_parameters.RemoveValue("x_coordinates");
-    test_parameters.AddVector("x_coordinates", ScalarVector{5, 1.0});
+    test_parameters.AddVector("x_coordinates", ScalarVector{2, 1.0});
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp:139]: Does this mean that if we create a 'constant' multiline process, with e.g. 2 tables in the parameters, we cannot have 5 x_coordinates (while the tables aren't used)? Please correct me if I'm wrong

**Review by rfaasse — COMMENTED** [review]
Nice catch and well tested and scoped fix! I have some questions left, mainly about the 'constant' version of the multiline process and the defaults.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp (hunk 5 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp:139]: There should be as many entries in the table list as there are points. However a entry in the table list may be 0 to indicate a constant value.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp (hunk 2 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_constant_phreatic_multi_line_pressure_process.cpp]: Indeed, reverted the wrong addition.

File: applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp (hunk 1 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp:55]: The points are chosen such that over a small domain they form a possible phreatic surface. ( if the gravity direction is in the xy plane )

File: applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp (hunk 3 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp:59]: Lets look at that together.

Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_processes/apply_constant_phreatic_multi_line_pressure_process.cpp:59]: Tried together, but without succes. The constructor of the "constant" process fails when it has no table in its defaults.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_phreatic_multi_line_pressure_table_process.cpp (hunk 4 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_phreatic_multi_line_pressure_table_process.cpp]: Added a test that does not throw

**Review by rfaasse — APPROVED** [review]
This is a nice sanity check, thanks for adding it! After our discussions, this is ready to go from my perspective

**Comment by rfaasse** [comment]
FYI We need to bring it up to date with master before we can merge (due to the failing windows pipeline being fixed)

**Pull request merged**